### PR TITLE
KOGITO-2469: implement synchronous mode in decision tracing addon

### PR DESCRIPTION
**Pull request by kostola**
- [JIRA task](https://issues.redhat.com/browse/KOGITO-2469)

## Goal
Extend the decision tracing addon with the capability to switch to a full synchronous mode in case tracing itself is considered critial and a possible increase in the API roundtrip time can be tolerated to make sure the events are completely processed before sending out the response.

This feature can be enabled/disabled with the `kogito.addon.tracing.decision.asyncEnabled` property. By default its value is `true`, which means the asyncronous mode is the default one.

### CODE REVIEW

File: addons/tracing/tracing-decision-springboot-addon/src/main/java/org/kie/kogito/tracing/decision/SpringBootDecisionTracingConfiguration.java [hunk 1]
```diff
@@ -55,6 +58,22 @@ public SpringBootDecisionTracingConfiguration(
...
+        return new SpringBootDecisionTracingCollectorAsync(eventEmitter, configBean, application);
+    }
+
+    @Bean
+    @ConditionalOnProperty(value = "kogito.addon.tracing.decision.asyncEnabled", havingValue = "false")
+    public SpringBootDecisionTracingCollector collectorForSyncMode(final SpringBootTraceEventEmitter eventEmitter,
+                                                                   final ConfigBean configBean,
+                                                                   final Application application) {
+        return new SpringBootDecisionTracingCollectorAsync(eventEmitter, configBean, application);
+    }
```
Comment (danielezonca) [inline addons/tracing/tracing-decision-springboot-addon/src/main/java/org/kie/kogito/tracing/decision/SpringBootDecisionTracingConfiguration.java:75]: Note 1: both are returning `SpringBootDecisionTracingCollectorAsync` instance
Note 2: what about adding the property as method parameter to have a single method? In this way we can prevent wrong situations like generate both beans if we will change parameter logic. Then please add a test to verify this producer method
```suggestion
    @Bean
    public SpringBootDecisionTracingCollector collectorForAsyncMode(final SpringBootTraceEventEmitter eventEmitter,
                                                                    final ConfigBean configBean,
                                                                    final Application application,
                                                                    @Value(value = "${kogito.addon.tracing.decision.asyncEnabled:true}") final boolean asyncEnabled) {
        if (asyncEnabled) {
            return new SpringBootDecisionTracingCollectorAsync(eventEmitter, configBean, application);
        } else {
            return new SpringBootDecisionTracingCollector(eventEmitter, configBean, application);
        }
    }
```

Comment (kostola) [reply addons/tracing/tracing-decision-springboot-addon/src/main/java/org/kie/kogito/tracing/decision/SpringBootDecisionTracingConfiguration.java]: Thanks for the catch and the suggestion. It's way better now.

Comment (danielezonca) [reply addons/tracing/tracing-decision-springboot-addon/src/main/java/org/kie/kogito/tracing/decision/SpringBootDecisionTracingConfiguration.java]: Can you please add a test for this method?

Comment (kostola) [reply addons/tracing/tracing-decision-springboot-addon/src/main/java/org/kie/kogito/tracing/decision/SpringBootDecisionTracingConfiguration.java]: Done :+1:

**Review by danielezonca — APPROVED** [review]
LGTM :+1:

File: addons/tracing/tracing-decision-quarkus-addon/src/test/java/org/kie/kogito/tracing/decision/QuarkusDecisionTracingTest.java [hunk 2]
```diff
@@ -35,76 +36,114 @@
...
     private static final String TEST_SERVICE_URL = "localhost:8080";
 
+    private static final ObjectMapper MAPPER = new ObjectMapper()
+            .registerModule(JsonFormat.getCloudEventJacksonModule());
+
     @Test
-    public void test_ListenerAndCollector_UseRealEvents_Working() throws IOException {
-        final DMNRuntime runtime = DMNKogito.createGenericDMNRuntime(new java.io.InputStreamReader(
-                QuarkusDecisionTracingTest.class.getResourceAsStream(MODEL_RESOURCE)
+    void test_AsyncListenerAndCollector_UseRealEvents_Working() throws IOException {
```
Comment (jiripetrlik) [inline addons/tracing/tracing-decision-quarkus-addon/src/test/java/org/kie/kogito/tracing/decision/QuarkusDecisionTracingTest.java:78]: Can we avoid underscores in the name of methods?

File: addons/tracing/tracing-decision-quarkus-addon/src/test/java/org/kie/kogito/tracing/decision/QuarkusDecisionTracingTest.java [hunk 3]
```diff
@@ -35,76 +36,114 @@
...
-        QuarkusDecisionTracingCollector collector = new QuarkusDecisionTracingCollector(mockedApplication, eventEmitter, configBean);
+        QuarkusDecisionTracingCollector collector = new QuarkusDecisionTracingCollector(eventEmitter, configBean, mockedApplication);
         eventEmitter.getEventPublisher().subscribe(subscriber);
-        eventCaptor.getAllValues().forEach(collector::onEvent);
+        events.forEach(collector::onEvent);
 
         subscriber.assertValueCount(1);
 
-        CloudEvent cloudEvent = MAPPER.readValue(subscriber.values().get(0), CloudEvent.class);
+        CloudEvent cloudEvent = CloudEventUtils.decode(subscriber.values().get(0)).orElseThrow(IllegalStateException::new);
```
Comment (jiripetrlik) [inline addons/tracing/tracing-decision-quarkus-addon/src/test/java/org/kie/kogito/tracing/decision/QuarkusDecisionTracingTest.java:143]: Would it be possible to add some message with description into the Exception.

File: addons/tracing/tracing-decision-springboot-addon/src/test/java/org/kie/kogito/tracing/decision/SpringBootDecisionTracingTest.java [hunk 4]
```diff
@@ -41,73 +43,113 @@
...
-        verify(template).send(topicCaptor.capture(), payloadCaptor.capture());
+        SpringBootDecisionTracingCollector collector = new SpringBootDecisionTracingCollector(eventEmitter, configBean, mockedApplication);
+        events.forEach(collector::onApplicationEvent);
 
-        assertEquals(TEST_TOPIC, topicCaptor.getValue());
+        ArgumentCaptor<String> payloadCaptor = ArgumentCaptor.forClass(String.class);
+        verify(mockedTemplate).send(eq(TEST_KAFKA_TOPIC), payloadCaptor.capture());
 
-        CloudEvent cloudEvent = OBJECT_MAPPER.readValue(payloadCaptor.getValue(), CloudEvent.class);
+        CloudEvent cloudEvent = CloudEventUtils.decode(payloadCaptor.getValue()).orElseThrow(IllegalStateException::new);
```
Comment (jiripetrlik) [inline addons/tracing/tracing-decision-springboot-addon/src/test/java/org/kie/kogito/tracing/decision/SpringBootDecisionTracingTest.java:147]: Can we add some message into the exception?

**Review by jiripetrlik — COMMENTED** [review]
@kostola
Thank you for your PR! I have few comments/suggestions.

File: addons/tracing/tracing-decision-quarkus-addon/src/test/java/org/kie/kogito/tracing/decision/QuarkusDecisionTracingTest.java (hunk 2 above)
Comment (kostola) [reply addons/tracing/tracing-decision-quarkus-addon/src/test/java/org/kie/kogito/tracing/decision/QuarkusDecisionTracingTest.java]: Removed all the underscores from test method names :+1:

File: addons/tracing/tracing-decision-quarkus-addon/src/test/java/org/kie/kogito/tracing/decision/QuarkusDecisionTracingTest.java (hunk 3 above)
Comment (kostola) [reply addons/tracing/tracing-decision-quarkus-addon/src/test/java/org/kie/kogito/tracing/decision/QuarkusDecisionTracingTest.java]: Done :+1:

File: addons/tracing/tracing-decision-springboot-addon/src/test/java/org/kie/kogito/tracing/decision/SpringBootDecisionTracingTest.java (hunk 4 above)
Comment (kostola) [reply addons/tracing/tracing-decision-springboot-addon/src/test/java/org/kie/kogito/tracing/decision/SpringBootDecisionTracingTest.java]: Done :+1:

**Review by jiripetrlik — APPROVED** [review]
@kostola 
Thank you for your changes! It looks good to me.

**Pull request merged**